### PR TITLE
Make `PauseTiming()` and `ResumeTiming()` per thread.

### PR DESCRIPTION
**Pull request by EricWF**
Currently we time benchmarks using a single global timer that tracks per-process CPU usage. Pausing and resuming this timer have to act as a barrier to all threads. This has crippling effects on multi-threaded benchmarks. If you pause every iterator you synchronize the entire benchmark. It's effectively no longer multi-threaded.

This patch changes to a per-thread timer. Instead of measuring process CPU time we sum thread CPU time and we pause on a per-thread basis. 

Below are comparison of the new and old results from `basic_test.cc`. Note that the `BM_spin_pause_during` test get 95% faster.

```
Benchmark                                                    Time           CPU
-------------------------------------------------------------------------------
BM_empty_mean                                               +0.00         +0.00
BM_empty/threads:4_mean                                     +0.00         +0.00
BM_spin_empty/8_mean                                        +0.00         +0.00
BM_spin_empty/512_mean                                      +0.01         +0.00
BM_spin_empty/8k_mean                                       -0.00         -0.00
BM_spin_empty/8/threads:4_mean                              +0.00         -0.17
BM_spin_empty/512/threads:4_mean                            +0.01         +0.00
BM_spin_empty/8k/threads:4_mean                             -0.01         +0.00
BM_spin_pause_before/8_mean                                 +0.00         +0.00
BM_spin_pause_before/512_mean                               +0.03         +0.02
BM_spin_pause_before/8k_mean                                +0.01         +0.01
BM_spin_pause_before/8/threads:4_mean                       +0.00         +0.00
BM_spin_pause_before/512/threads:4_mean                     +0.04         +0.01
BM_spin_pause_before/8k/threads:4_mean                      -0.03         -0.00
BM_spin_pause_during/8_mean                                 -0.24         -0.25
BM_spin_pause_during/512_mean                               -0.24         -0.24
BM_spin_pause_during/8k_mean                                -0.13         -0.13
BM_spin_pause_during/8/threads:4_mean                       -0.97         -0.90
BM_spin_pause_during/512/threads:4_mean                     -0.96         -0.89
BM_spin_pause_during/8k/threads:4_mean                      -0.95         -0.85
BM_pause_during_mean                                        -0.23         -0.20
BM_pause_during/threads:4_mean                              -0.97         -0.90
BM_pause_during/real_time_mean                              -0.24         -0.26
BM_pause_during/real_time/threads:4_mean                    -0.97         -0.90
BM_spin_pause_after/8_mean                                  +0.00         +0.00
BM_spin_pause_after/512_mean                                +0.00         +0.00
BM_spin_pause_after/8k_mean                                 -0.00         -0.00
BM_spin_pause_after/8/threads:4_mean                        +0.00         +0.00
BM_spin_pause_after/512/threads:4_mean                      -0.01         -0.02
BM_spin_pause_after/8k/threads:4_mean                       +0.01         +0.01
BM_spin_pause_before_and_after/8_mean                       +0.00         +0.00
BM_spin_pause_before_and_after/512_mean                     +0.00         +0.00
BM_spin_pause_before_and_after/8k_mean                      +0.01         +0.01
BM_spin_pause_before_and_after/8/threads:4_mean             +0.00         +0.00
BM_spin_pause_before_and_after/512/threads:4_mean           +0.06         +0.04
BM_spin_pause_before_and_after/8k/threads:4_mean            -0.00         +0.02
BM_empty_stop_start_mean                                    +0.00         +0.00
BM_empty_stop_start/threads:4_mean                          +0.00         +0.00
```

There's still work to do on this, but I was hoping for initial feedback on the direction.

### CODE REVIEW

File: src/benchmark.cc [hunk 1]
```diff
@@ -861,19 +797,17 @@ RunBenchmark(const benchmark::internal::Benchmark::Instance& b,
             thread.join();
         }
         for (std::size_t ti = 0; ti < pool.size(); ++ti) {
-            pool[ti] = std::thread(&RunInThread, &b, iters, static_cast<int>(ti), &total);
+            pool[ti] = std::thread(&RunInThread, &b, iters,
+                                   static_cast<int>(ti + 1), &manager);
```
Comment (dominichamon) [inline src/benchmark.cc]: is this an existing bug or are you incrementing thread_id for some reason related to the PR?

File: src/cpu_thread_time.cc [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  struct rusage ru;
+  if (getrusage(RUSAGE_CHILDREN, &ru) == 0) {
+    return MakeTime(ru);
+  } else {
+    return 0.0;
+  }
+#else
+  // TODO: Not sure what this even means on Windows
+  return 0.0;
+#endif  // OS_WINDOWS
```
Comment (dominichamon) [inline src/cpu_thread_time.cc]: nit: BENCHMARK_OS_WINDOWS

**Comment by dominichamon** [comment]
the direction looks great. it's very clean and clearly more correct.

File: src/benchmark.cc (hunk 1 above)
Comment (EricWF) [inline src/benchmark.cc]: I changed it so the main thread always runs as thread `0` for multi-threaded benchmarks. Previously we just slept the main thread.

**Comment by EricWF** [comment]
Unfortunately this patch is starting to get big. I had to fix a bunch of TSAN race conditions, including existing ones in CHECK and VLOG.

However I have managed to remove all (?) of the global state used to run the benchmarks.

File: src/benchmark.cc [hunk 3]
```diff
@@ -120,148 +109,118 @@ static const int kRangeMultiplier = 8;
...
+public:
+    ThreadManager(int num_threads)
+        : alive_threads_(num_threads),
+          start_stop_barrier_(num_threads)
+    {
+    }
 
-// TODO(ericwf): support MallocCounter.
-//static benchmark::MallocCounter *benchmark_mc;
+    Mutex& getBenchmarkMutex() const
```
Comment (dominichamon) [inline src/benchmark.cc]: can you do a format/style pass at some point? GetBenchmarkMutex, etc.

Comment (EricWF) [inline src/benchmark.cc]: I'll do it right now if you clarify what the style should be (Other than FunctionNameAllUpper).

Comment (dominichamon) [inline src/benchmark.cc]: i think that's all it is.. clang-format Google-style should take care of the rest (there's some extra indents around the place).

Comment (EricWF) [inline src/benchmark.cc]: Can we check in a .clang-format file to make formatting easier?

Comment (dominichamon) [inline src/benchmark.cc]: done :)

**Comment by dominichamon** [comment]
this lgtm. have you run a comparison for all the benchmark tests to see how the timing changes?

**Comment by EricWF** [comment]
> this lgtm. have you run a comparison for all the benchmark tests to see how the timing changes?

I've done some. The results for `basic_test.cc` are at the top of this PR. I just need to test this a little more on OS X which uses a really weird implementation.

**Comment by EricWF** [comment]
>  have you run a comparison for all the benchmark tests to see how the timing changes?

I have now. The results are very close to the same except for on threaded benchmarks. Surprisingly threaded benchmarks that *don't* use PauseTiming still benefit, sometimes greatly.

Previously all threads would have to complete the KeepRunning loop before the timer could be paused, this essentially made each benchmark as slow as its slowest thread. This added as much as 30% to the real time measurements.  Now that has been fixed.

**Comment by dominichamon** [comment]
wonderful!

**Pull request merged**